File: climbing-stairs/solution70.py
# https://leetcode.com/problems/climbing-stairs/

class Solution70:

    # T: O(n^2), S: O(n)
    def climbStairsRec(self, n: int) -> int:
        if n < 3:
            return n
        return self.climbStairsRec(n - 1) + self.climbStairsRec(n - 2)

    # T: O(n), S: O(n)
    def climbStairsMemo(self, n: int) -> int:
        memo = [0] * (n + 1)
        return self.helperMemo(n, memo)

    def helperMemo(self, n: int, memo: list) -> int:
        if memo[n] > 0:
            return n
        if n < 3:
            memo[n] = n
            return n
        memo[n] = self.helperMemo(n - 1, memo) + self.helperMemo(n - 2, memo)
        return memo[n]

    # T: O(n), S: O(n)
    def climbStairsDp(self, n: int) -> int:
        if n == 0 or n == 1:
            return n
        dp = [0] * (n + 1)
        dp[1] = 1
        dp[2] = 2
        for i in range(3, n + 1):
            dp[i] = dp[i - 1] + dp[i - 2]
        return dp[n]

    # T: O(n), S: O(1)
    def climbStairsDpBottomUpOpt(self, n: int) -> int:
        if n < 3:
            return n
        a, b, sum = 1, 2, 0
        for i in range(3, n + 1):
            sum = a + b
            a = b
            b = sum
        return sum


def main() -> None:
    sol = Solution70()
    print(sol.climbStairsRec(3))  # 3
    print(sol.climbStairsMemo(3))  # 3
    print(sol.climbStairsDp(3))  # 3
    print(sol.climbStairsDpBottomUpOpt(3))  # 3
    print(sol.climbStairsRec(4))  # 5
    print(sol.climbStairsMemo(4))  # 5
    print(sol.climbStairsDp(4))  # 5
    print(sol.climbStairsDpBottomUpOpt(4))  # 5


if __name__ == "__main__":
    main()

File: climbing-stairs/Solution70.java
// https://leetcode.com/problems/climbing-stairs/

class Solution70 {

    // T: O(n), S: O(1)
    public int climbStairs(int n) {
        if (n < 3) {
            return n;
        }
        int a = 1, b = 2, sum = 0;
        for (int i = 3; i <= n; i++) {
            sum = a + b;
            a = b;
            b = sum;
        }
        return sum;
    }

    // T: O(n), S: O(n)
    public int climbStairsDp(int n) {
        if (n == 0 || n == 1) {
            return n;
        }
        var dp = new int[n + 1];
        dp[1] = 1;
        dp[2] = 2;
        for (int i = 3; i <= n; i++) {
            dp[i] = dp[i - 1] + dp[i - 2];
        }
        return dp[n];
    }

    // T: O(n), S: O(n)
    public int climbStairsMemo(int n) {
        var memo = new int[n + 1];
        return helperMemo(0, n, memo);
    }

    private int helperMemo(int i, int n, int[] memo) {
        if (i > n) {
            return 0;
        }
        if (i == n) {
            return 1;
        }
        if (memo[i] > 0) {
            return memo[i];
        }
        memo[i] = helperMemo(i + 1, n, memo) + helperMemo(i + 2, n, memo);
        return memo[i];
    }

    // T: O(n^2), S: O(n)
    public int climbStairsBrute(int n) {
        if (n < 3) {
            return n;
        }
        return climbStairs(n - 1) + climbStairs(n - 2);
    }

    public static void main(String[] args) {
        var sol = new Solution70();
        System.out.println(sol.climbStairs(3)); // 3
        System.out.println(sol.climbStairsDp(3));
        System.out.println(sol.climbStairsMemo(3));
        System.out.println(sol.climbStairsBrute(3));
        System.out.println(sol.climbStairs(4)); // 5
        System.out.println(sol.climbStairsDp(4)); // 5
        System.out.println(sol.climbStairsMemo(4)); // 5
        System.out.println(sol.climbStairsBrute(4)); // 5
    }

}
